src/redprl/refiner_types.fun

```diff
@@ -53,36 +53,25 @@ struct
    * for example: it has elements `base`, `loop` and `fcom`; there are infinitely
    * many different shapes of `fcom` and thus infinitely many cases (unless our LF
    * can deal with them uniformly), and there are (infinitely) many equalities
-   * between these elements due to cubicality.
+   * between these elements due to cubicality. Another example is the natural
+   * number type `nat` which has elements `zero`, `(succ zero)`, ...; this also
+   * leads to infinitely many cases.
    *
-   * Another example is the natural number type `nat` which has elements `zero`,
-   * `(succ zero)`, ...; this also has infinitely many cases. It is imaginable
-   * that we can generate two "higher-order subgoals" in some higher-order LF:
-   *
-   * 1. `\Gamma, x:nat, \Delta[zero/x] >> J[zero/x]`
-   * 2. if `\Gamma, x:nat, \Delta[#n/x] >> J[#n/x]` is provable for some
-   *    numeral `n`, then `\Gamma, x:nat, \Delta[(succ #n)/x] >> J[(succ #n)/x]`
-   *    is provable.
-   *
-   * One also has to think carefully about how to construct proper realizers.
-   *
-   *
-   * Before we have a more powerful LF, however, we can still have a "weaker" or
-   * "local" elimination principle, `Elim`, that is just trying to generate a
-   * term that would satisfy the structural equality of the eliminator (EqElim).
-   * For the natural number type, the weaker elimination principle generates
-   * the following two subgoals:
+   * Before we have a more powerful LF, however, we can still have the second
+   * category of elimination rules, `Elim`, which is just trying to generate
+   * a term that would satisfy the structural equality (EqElim). For the natural
+   * number type, the elimination rule of this category generates these two subgoals
    *
    * 1. `\Gamma, x:nat, \Delta >> C[zero/x] true`
    * 2. `\Gamma, x:nat, \Delta, y:nat, z:C[y/x] >> C[(succ y)/x] true`
    *
    * and combine the realizers from the two subgoals into something of the form
    * `(nat-rec x ...)`.
    *
-   * There is no reason to have a weaker elimination principle if the stronger
-   * one is already there. Therefore, there is `Bool.FullElim` but not `Bool.Elim`.
-   * In Nuprl, everything is named "elimination" but I (favonia) think maybe
-   * it is useful to remind tactic writers of what is really going on.
+   * There is no reason to have an elimination principle of the second category
+   * if there exists one of the first category. Therefore, there is `Bool.FullElim`
+   * but no `Bool.Elim`. In Nuprl, everything is named "elimination" but I (favonia)
+   * think it might be useful to remind tactic writers of its category.
```

Comment:
> ... but I do not see how to deal with other kinds of goals by introducing hypotheses, and perhaps they are closely related issues?

Could you elaborate? You had brought up questions about doing things like matching on things like fcoms, but I think that's a separate matter and IMO we shouldn't think about it for the time being. My preference is if we never add facilities for doing that, though maybe some day it will become necessary. In any case, since my default position is that we wouldn't do this, I don't want to have a lot of comments and documentation that talk about it, since it may make things confusing for people as they try to learn from the codebase.

> Did you mean the former kind?

Oops, yes!